flexmeasures/data/utils.py

```diff
@@ -44,3 +49,112 @@ def get_data_source(
+
+    :param data: BeliefsDataFrame (or a list thereof) to be saved
+    :param save_changed_beliefs_only: if True, beliefs that are already stored in the database with an earlier belief time are dropped.
+    :returns: status string (or a list thereof), one of the following:
+              - 'success': all beliefs were saved
+              - 'success_with_replacements': all beliefs were saves, (some) replacing pre-existing beliefs
+              - 'success_but_data_empty': there was nothing to save
+              - 'success_but_nothing_new': no beliefs represented a state change
+              - 'success_but_partially_new': not all beliefs represented a state change
+              - 'failed_due_to_forbidden_replacements': no beliefs were saved, because replacing pre-existing beliefs is forbidden
+    """
+
+    # Convert to list
+    if not isinstance(data, list):
+        timed_values_list = [data]
+    else:
+        timed_values_list = data
+
+    success_list = []
+    for timed_values in timed_values_list:
+
+        if timed_values.empty:
+            # Nothing to save
+            success_list.append("success_but_data_empty")
+            continue
+
+        len_before = len(timed_values)
+        if save_changed_beliefs_only:
+
+            # Drop beliefs that haven't changed
+            timed_values = (
+                timed_values.convert_index_from_belief_horizon_to_time()
+                .groupby(level=["belief_time", "source"], as_index=False)
+                .apply(drop_unchanged_beliefs)
+            )
+            len_after = len(timed_values)
+
+            # Work around bug in which groupby still introduces an index level, even though we asked it not to
+            if None in timed_values.index.names:
+                timed_values.index = timed_values.index.droplevel(None)
+
+            if timed_values.empty:
+                # No state changes among the beliefs
+                success_list.append("success_but_nothing_new")
+                continue
+        else:
+            len_after = len_before
+
+        # if timed_values.empty or (save_changed_beliefs_only and len_after < len_before):
+        #     current_app.logger.info("Nothing new to save")
+        #     success_list.append(False)  # no data or data already existed or data doesn't represent updated beliefs
+        # else:
+        current_app.logger.info("SAVING TO DB...")
+        try:
+            TimedBelief.add_to_session(
+                session=db.session, beliefs_data_frame=timed_values
+            )
+            db.session.flush()
+            db.session.commit()
+            if len_after < len_before:
+                # new data was saved
+                success_list.append("success_but_partially_new")
+            else:
+                # all data was saved
+                success_list.append("success")
+        except IntegrityError as e:
+            current_app.logger.warning(e)
+            db.session.rollback()
```

Comment:
When we roll back, we also roll back everything we did before on entries within `timed_values_list`.

So at the very least `success_list = []` is in order, but probably we need to think some more about this. 

I don't think we should handle a list of BDFs and commit them separately. Per request, we should try to only have one transaction, one commit. 

(by the way, my code in data/transactional is not really used by us within the normal API code, more for CLI functions, besides the `after_request_exception_rollback_session` event listener which we registered with each request)

Reply:
I'll address this comment and your comment about committing together.

What did I do already? I changed the `save_to_db` function to handle the list of BeliefsDataFrames as a single transaction. I also found out that a commit is not necessary. A flush will suffice to bring out the `UniqueViolation` (which I now check for more explicitly).

My follow-up question is: given that the function may roll back the entire session, would it be wise to have this function start a session of its own? [With something like this.](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#framing-out-a-begin-commit-rollback-block)

Reply:
That might be a good path forward. However, one more simplifying thought, as it is still quite complex stuff:

Why not set `allow_overwrite=True` when in play mode, and be done with it? No need to check for `UniqueViolations`.

Reply:
Thanks, I simplified accordingly.

After some quick tests with opening a fresh session, I found that the objects we are saving (TimedBeliefs) link to other objects (Sensors and DataSources) bound to the previous session, which results in an informative SQLAlchemy error:
```
sqlalchemy.exc.InvalidRequestError: Object '<Sensor at 0x7fd3dfc6b040>' is already attached to session '1' (this is '2')
```
I now don't think this is a very fruitful path. I'll instead add a statement to the docstring that the `save_to_db` method may roll back the session, and advice to keep transactions short ([as recommended](https://docs.sqlalchemy.org/en/14/orm/session_basics.html#when-do-i-construct-a-session-when-do-i-commit-it-and-when-do-i-close-it)).

Reply:
Do we really need to check for `IntegrityError` at all? 

You now roll back all `IntegrityError`s, but re-raise `UniqueError`s. Still confusing.

Can you state the goal of that try/except block?

Note that we already roll back if an exception happened (in `after_request_exception_rollback_session`). 

What if we had no try/except block?
What if we only check for `UniqueError`s?

Reply:
> You now roll back all `IntegrityError`s, but re-raise `UniqueError`s. Still confusing.

That doesn't seem right:
- We're talking about `UniqueViolation`, not `UniqueError`.
- We don't re-raise `UniqueViolation`. It's the one exception we **don't** re-raise. Instead, we have a pretty good idea what happened, so we choose to report back to the user that it shouldn't try to replace existing beliefs.

> Note that we already roll back if an exception happened (in after_request_exception_rollback_session).
  What if we had no try/except block?

I think that would result in the user getting a 500 status code with a less informative message.

> What if we only check for UniqueErrors?

The `UniqueViolation` is (for lack of a better word) 'wrapped' by SQLAlchemy into an `IntegrityError`. Therefore, I except all `IntegrityError`s, then pick out the ones that were wrapping a `UniqueViolation` (using `e.orig`) and re-raise the ones that didn't. I don't know how to directly check for a `UniqueViolation`.

Reply:
If you prefer, I could move the rollback to after the re-raising if-block, so it only rolls back here in case of a `UniqueViolation`, and otherwise rolls back in `after_request_exception_rollback_session`.

Reply:
Sorry for being imprecise. Head's in a different PR right now.

I still need more clarity what is being achieved here and why. Also in a comment for other readers. 

As I understand it, we are trying to support simulations (which we call "play" mode), where overwriting existing data is allowed. Not quite sure why that is needed, either. Probably to lower the effort of trying different scenarios.

You do have a comment, but next to the reason why we might overwrite, I'm also not sure when that try/except block needs to be applied (now that there is no commit). Isn't it only relevant when not in play mode? 

Finally, the reason we do the try/except block seems to be to give an informative error. Is that true?

Reply:
> the reason we do the try/except block seems to be to give an informative error. Is that true?

Yes.

Let's discuss the other issues over a call.

Reply:
[Here](https://github.com/FlexMeasures/flexmeasures/pull/289/commits/fec17b734f975340194e7925f1cb76508671c371)'s an alternative using an error handler. Probably better to move it out of that `__init__.py`. To discuss.